### PR TITLE
Upstream sync 2012-04-13

### PR DESCRIPTION
**Pull request by dszmigielski**
### Rewrite Accurate_Duration test to reduce flakiness  ([#1360](https://github.com/DataDog/dd-trace-dotnet/pull/1360))
Author: Kevin Gosse <krix33@gmail.com>
Date:   Tue Apr 13 15:39:35 2021 +0200

### Don't run Unit test or runner pipelines on all branch pushes ([#1354](https://github.com/DataDog/dd-trace-dotnet/pull/1354))
commit 0a1701129d46e65c92ddb57bee4635c4a41f71f4 (upstream/master)
Author: Andrew Lock <andrew.lock@datadoghq.com>
Date:   Mon Apr 12 18:13:16 2021 +0100
    
    There's a lot of duplicated effort caused by these branch-only pushes, and with limited runners available (especially macos), this causes a backlog of runs for PRs
    
    We can always manually trigger a pipeline for a branch if you want a pre-PR check, so I think this is probably a good option?

### Try fix scheduled benchmark/throughput runs ([#1359](https://github.com/DataDog/dd-trace-dotnet/pull/1359))
commit 963acb3e11d6f2f01b7671a3f9197270d035c65a
Author: Andrew Lock <andrew.lock@datadoghq.com>
Date:   Mon Apr 12 18:12:58 2021 +0100

### Fix Datadog.Trace.Tests.DogStatsDTests.Send_metrics_when_enabled ([#1358](https://github.com/DataDog/dd-trace-dotnet/pull/1358))
commit e321e6827ac031c07b36090667a0ed3f45701e43
Author: Zach Montoya <zach.montoya@datadoghq.com>
Date:   Mon Apr 12 09:53:08 2021 -0700
    
    Fix two bad assumptions in the Statsd assertions:
    1. The heartbeat fires on a timer that is independent of sending a trace, so do not assert its presence
    2. Sending a trace may require multiple requests to the trace agent since requests may fail, so assert that it's fired 1+ times instead of exactly once

### Upgrade Serilog & Serilog.Sinks.File Vendors ([#1345](https://github.com/DataDog/dd-trace-dotnet/pull/1345))
commit 483c290a1c35993b8432e4aebcfedd92104dc741
Author: Colin Higgins <colin-higgins@users.noreply.github.com>
Date:   Mon Apr 12 12:47:39 2021 -0400

### Add benchmark for Serilog log injection ([#1351](https://github.com/DataDog/dd-trace-dotnet/pull/1351))
commit 9d421e72ff4814dc3dcceb416b2b46e841664eb2
Author: Kevin Gosse <krix33@gmail.com>
Date:   Mon Apr 12 16:31:08 2021 +0200

### Log a warning if partial flush is enabled with an old agent ([#1347](https://github.com/DataDog/dd-trace-dotnet/pull/1347))
commit e75fd566d8293e452e2c5bef3703ba8d3612f42d
Author: Kevin Gosse <krix33@gmail.com>
Date:   Mon Apr 12 11:28:32 2021 +0200

### Fix small sync over async occurrence in DatadogHttpClient ([#1348](https://github.com/DataDog/dd-trace-dotnet/pull/1348))
commit e2e1b3397c72e92f1ad0c3c6b8d8b9b70681bfe1
Author: Zach Montoya <zach.montoya@datadoghq.com>
Date:   Thu Apr 8 16:25:49 2021 -0700

    Co-authored-by: Kevin Gosse <krix33@gmail.com>

### Fix the package build pipeline ([#1356](https://github.com/DataDog/dd-trace-dotnet/pull/1356))
commit c2a588e0138f118573adad2a4e3604e1ad5c1e2b
Author: Zach Montoya <zach.montoya@datadoghq.com>
Date:   Thu Apr 8 09:55:36 2021 -0700

    This build break was inadvertently introduced in https://github.com/DataDog/dd-trace-dotnet/pull/1309 because the build configuration was not tweaked appropriately for the package scripts and pipeline.

### Fix native unit tests running in Release mode ([#1355](https://github.com/DataDog/dd-trace-dotnet/pull/1355))
commit f6a53dfd651025181025d462e7aada67ce67233c
Author: Andrew Lock <andrew.lock@datadoghq.com>
Date:   Thu Apr 8 15:52:21 2021 +0100

###  Update Uri cleaning algorithm to be more aggressive ([#1327](https://github.com/DataDog/dd-trace-dotnet/pull/1327))
commit e2a5e12548bb396f85c54ebf9782fe915eab6053
Author: Andrew Lock <andrew.lock@datadoghq.com>
Date:   Thu Apr 8 15:13:27 2021 +0100

    * Update Uri cleaning algorithm to be more aggressive
    
    Replaces any hex-ish (alpha or `-`) string with `?`.
    If it's numbers and dashes only, always replaces
    If it contains alpha, must be at least 16 chars long and contain a number
    Added in back-compat way, that avoids changing internal (public) APIs
    
    * Conditionally enable the new URI cleaning algorithm in ASP.NET
    
    Enabled when the new route names flag is enabled
    Added a note that we're currently using the _legacy_ algorithm still in the generic scope factory for _outbound_ requests. We may want to update this, but wasn't sure if we should do it under the same flag, as it's _kinda_ unrelated
    
    * Simplify implementation
    
    * Mark older UriHelpers methods Obsolete (kept for back compat)
    
    Add documentation to CleanUri method
    Update tests to use non-obsolete methods
    
    * Use new Uri cleaning in AspNetCore resources when feature flag enabled
    
    * Remove the feature-flag behaviour i.e. always use the new Id replacement algorithm
    
    We decided to take this approach as it's simpler, and in practice the cleaning algorithm should only be used when there's a 404 anyway. If customers find the new algorithm is causing them problems, opting in to the new route-based resource names should fix it.
    
    The one other place this cleaning algorithm is used is in the ScopeFactory for outbound HTTP requests. There's no workaround there, but the chances of issues are small, so we're ok with the tradeoff for simplicity and performance

### Simplify moving average calculator ([#1350](https://github.com/DataDog/dd-trace-dotnet/pull/1350))
commit 201247af4449d18077ca7a76894a02b1110be7b7
Author: Andrew Lock <andrew.lock@datadoghq.com>
Date:   Thu Apr 8 09:36:26 2021 +0100
    
    * Simplify moving average calculator
    
    Was previously going a bit overboard with keeping things atomic, when in reality we only need the current buckets and keep rate to be atomic
    
    * Update src/Datadog.Trace/Agent/MovingAverageKeepRateCalculator.cs
    
    Co-authored-by: Lucas Pimentel-Ordyna <lucas.pimentel@datadoghq.com>
    
    * Apply Kevin-level optimisations
    
    Co-authored-by: Kevin Gosse <krix33@gmail.com>
    
    Co-authored-by: Lucas Pimentel-Ordyna <lucas.pimentel@datadoghq.com>
    Co-authored-by: Kevin Gosse <krix33@gmail.com>

### Fix Datadog.Trace.ClrProfiler.Native.sh ([#1309](https://github.com/DataDog/dd-trace-dotnet/pull/1309))
commit 80b4de8d039f728295ec4344bc25486ea9b4e3f6
Author: Zach Montoya <zach.montoya@datadoghq.com>
Date:   Wed Apr 7 13:33:53 2021 -0700
    
    Port changes from OTEL PR: https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/pull/77
    
    Adds `buildConfiguration` and `publishTargetFramework` variables into shell scripts

### Downgrading .NET SDK on Crank pipeline ([#1352](https://github.com/DataDog/dd-trace-dotnet/pull/1352))
commit 9da60a8e21c5f64a7f6157119d228f0b99436464
Author: Tony Redondo <tony.redondo@datadoghq.com>
Date:   Wed Apr 7 20:16:08 2021 +0200

### OTel Port: Fix double parsing
commit 58dd1b02dfb8e90923a30513b1fbc6b9cee9b0d2
Author: Zach Montoya <zach.montoya@datadoghq.com>
Date:   Wed Apr 7 09:03:04 2021 -0700
    
    Source: https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/pull/96 (#1349)

### Fix references to dotnet-tracer-native.log ([#1329](https://github.com/DataDog/dd-trace-dotnet/pull/1329))
commit be32faeaebef8664da0c7e9be23f47cac3cb1d10
Author: Andrew Lock <andrew.lock@datadoghq.com>
Date:   Wed Apr 7 12:28:50 2021 +0100
    
    * Copy all native log files
    
    With rolling native log files, there's more than one log file, and we don't know its name ahead of time, so use globbing
    
    * Replace obsolete DD_TRACE_LOG_PATH -> DD_TRACE_LOG_DIRECTORY
    
    * We now produce multiple native log files, so need to publish the directory
    
    * No need to touch file that won't be used
    
    * Try outputting log files after execution, as can't easily tail them
    
    * Also publish coredumps
    
    * Add EOLs
    
    * Use trap where possible
    
    * Reenable -e
    
    For simplicity, reinstate fail on first error, and use trap to copy logs and crashdumps
    
    * Fix environment variable
    
    Co-authored-by: Zach Montoya <zach.montoya@datadoghq.com>
    
    Co-authored-by: Zach Montoya <zach.montoya@datadoghq.com>

### Increase diagnostic observer test timeout ([#1331](https://github.com/DataDog/dd-trace-dotnet/pull/1331))
commit b3319d1775621054494f59aae41c9c3466e7abc4
Author: Andrew Lock <andrew.lock@datadoghq.com>
Date:   Wed Apr 7 10:02:03 2021 +0100]

### Partial OTEL port to improve TagsList ([#1311](https://github.com/DataDog/dd-trace-dotnet/pull/1311))
commit 18e6578c9777d07bf6802b59c7e379ca516fd20b
Author: Zach Montoya <zach.montoya@datadoghq.com>
Date:   Tue Apr 6 17:41:52 2021 -0700
    
    I looked to port https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/pull/71
    
    However, I don't think we necessarily need to take the changes from there, which can be summed up as adding a Metrics and Tags property on the ITags interface, to allow their Zipkin exporter to write the tags/metrics.
    
    In the end, I just ported one unit test from the PR and modified it to work with our structure.

### Improve DD_TRACE_HEADER_TAGS to decorate web server spans based on response headers ([#1301](https://github.com/DataDog/dd-trace-dotnet/pull/1301))
commit 04e76265d658acdb861cd20012d6339edecd2861
Author: Zach Montoya <zach.montoya@datadoghq.com>
Date:   Tue Apr 6 16:30:47 2021 -0700
    
    Previously, the `DD_TRACE_HEADER_TAGS` setting would accept a list of `<http_header_name>:<tag_name>` pairs and add tags to web server spans with the specified tag name if the header was present on an incoming HTTP request. This PR adds functionality to also consider headers from the HTTP response the web server issues. Additional changes include:
    
    1. In accordance with the HTTP Headers Tracing RFC, `DD_TRACE_HEADER_TAGS` now accepts entries without a specified tag name. In this scenario, the resulting tag name will be `http.<request|response>.headers.<normalized_header_name>`.
    2. All header tag names are now being normalized
    
    Affected integrations:
    - AspNet
    - AspNetMvc
    - AspNetWebApi2
    - AspNetCore
    
    Testing:
    - Local testing for all integrations
    - AspNetCore integration test added

### Add AssemblyLoadContext and Assembly.LoadFile testing to ServiceStackRedis sample ([#1337](https://github.com/DataDog/dd-trace-dotnet/pull/1337))
commit 018424d73b0a53eca52a0de8d9654ab06503229f
Author: Zach Montoya <zach.montoya@datadoghq.com>
Date:   Tue Apr 6 09:34:07 2021 -0700
    
    To further test resiliency against assembly loading issues, add AssemblyLoadContext and Assembly.LoadFile testing to the ServiceStackRedis sample. This augments the existing ADO.NET tests that do a similar thing, except in the ServiceStack.Redis library, all of the types are defined in the third-party library, unlike ADO.NET where some types are defined in the framework-provided System.Data assembly.

### Replace Task.Delay in dogstatsd with Thread.Sleep ([#1344](https://github.com/DataDog/dd-trace-dotnet/pull/1344))
commit 3fed1f37b5f503cd4f57424f56ac519632f33275
Author: Colin Higgins <colin-higgins@users.noreply.github.com>
Date:   Tue Apr 6 09:30:44 2021 -0400

### Report Trace Drop/Keep Rate to the Agent ([#1306](https://github.com/DataDog/dd-trace-dotnet/pull/1306))
commit af85a07a5a95a2f1466545ad6c1e3f0739d47328
Author: Andrew Lock <andrew.lock@datadoghq.com>
Date:   Tue Apr 6 10:57:43 2021 +0100
    
    * Add a moving average keep rate calculator
    
    Uses a simple moving average implementation to calculate the rate at which traces were kept/dropped
    Handles up to uint.MaxValue traces in a given timespan (1s by default, averaged over 10s)
    
    * Add _dd.tracer_kr metric to root spans of traces
    
    Records the current tracer keep rate on the root span of traces.
    
    * Switch to TaskCompletionSource instead of CancellationToken
    
    * Make _dd.trace_kr internal
    
    * Apply suggestions from code review
    
    Co-authored-by: Lucas Pimentel-Ordyna <lucas.pimentel@datadoghq.com>
    
    * Rename _keepRateCalculator -> _traceKeepRateCalculator for clarity
    
    * Update comment based on PR feedback
    
    * Fix variable name
    
    * Cast to CommonTags and use property where possible
    
    * Fix the windowSize named parameter
    
    * Fix error message
    
    * Reduce allocations in the UpdateBucketTaskLoopAsync task
    
    * Remove unused variable
    
    * Fix references
    
    * Add missing variable
    
    Co-authored-by: Lucas Pimentel-Ordyna <lucas.pimentel@datadoghq.com>

### Update ASP.NET Core Resource Names ([#1289](https://github.com/DataDog/dd-trace-dotnet/pull/1289))
commit a001b21b9276adef17dadd94911e12ebd0fb7b0a
Author: Andrew Lock <andrew.lock@datadoghq.com>
Date:   Tue Apr 6 10:21:03 2021 +0100
    
    * Add route template handling for AspNetCore diagnostic Observer
    
    * Add tags for ASP.NET Core MVC actions
    
    * Add Endpoint routing support to AspNetCoreDiagnosticObserver
    
    Adds the AspNetEndpoint tag
    Required adding a FrameworkReference which is likely not workable in the longer term.
    
    * Added support for tagging Razor Page in AspNetCoreDiagnosticObserver
    
    * Improve AspNetCoreDiagnosticObserver
    
    Reduce string allocations
    Fix pipeline re-execution behaviour
    
    * Add FeatureFlag for controlling new resource names in ASP.NET Core
    
    * Add tests for new diagnostic source behaviour
    
    * Convert ASP.NET Core endpoint-routing support to duck typing
    
    * Enable endpoint routing support for netcoreapp3.0
    
    * Significant rework of the ASP.NET Core diagnostic observer
    
    Create a separate span (aspnet_core.mvc) for each MVC execution
    This is more complex than it initially seems due to the use of endpoint routing.
    To more easily pass values between the various listener endpoints, makes use of the HttpContext.Features collection, to pass a feature through.
    Additionally:
     - Parent span resource name is updated based on route name
     - Parent span tagged with the _first_ selected endpoint and route
     - Child (MVC) spans contains action/controller etc tags + same route as parent.
     - In error/pipeline re-execution scenarios, the parent span route is not updated, but you may see additional MVC spans (depending on pipeline configuration)
    
     All this is controlled by a feature flag - with the feature flag off (the default) the resource names will remain the same as currently shipped
    
     Updated the tests as appropriate, including checking child spans. Could do with refactoring (e.g. snapshot testing of spans)
    
    * Create separate tags for ASP.NET Core
    
    Instead of re-using the ASP.NET ones
    Uses the pattern `aspnet_core.TAG`
    
    * Remove unnecessary use of nested StringBuilder
    
    * Fix case where IEndpointFeature is explicitly implemented
    
    This occurs with Kestrel for example, where Http2StreamofT implements HttpProtocol, which explicitly implements IEndpointFeature. Our duck typing library doesn't currently catch that case, so this approach uses a method duck type instead
    
    * Improve duck typing a bit
    
    * Refactor RequestTrackingFeature in aspnetcore diagnostic observer
    
    Handles cases where the pipeline is reexecuted before the endpoint routing executes. Occurs in cases such as a 404 handled by the statuscodepageswithrexecute handler, or errors in a "map" pipeline that are handled by the exception handler with reexecution
    
    * Fix copy pasta
    
    * Use StringBuilderCache in AspNetCoreDiagnosticObserver
    
    * Rename Feature Flag to DD_TRACE_ROUTE_TEMPLATE_RESOURCE_NAMES_ENABLED
    
    * Add parameter for readability
    
    Co-authored-by: Lucas Pimentel-Ordyna <lucas.pimentel@datadoghq.com>
    
    * Reduce number of ToLowerInvariant() calls when simplifying route pattern
    
    * Rename aspnet_core.mvc -> aspnet_core_mvc.request
    
    * Rename RequestTrackingFeature.Url -> OriginalUrl
    
    * Fix Comment
    
    Co-authored-by: Zach Montoya <zach.montoya@datadoghq.com>
    
    * Guard against null EndpointFeatureType
    
    * Add doc comments to duck typed proxies
    
    * Reorder methods to match order they are called at runtime
    
    Co-authored-by: Lucas Pimentel-Ordyna <lucas.pimentel@datadoghq.com>
    Co-authored-by: Zach Montoya <zach.montoya@datadoghq.com>

### Delete OrleansCrash and remove it from our integration test suite ([#1271](https://github.com/DataDog/dd-trace-dotnet/pull/1271))
commit f4c95b717f045ec06432642609386511da7e8286
Author: Zach Montoya <zach.montoya@datadoghq.com>
Date:   Mon Apr 5 14:38:18 2021 -0700
    
    This sample was originally added in https://github.com/DataDog/dd-trace-dotnet/pull/443 to fix an issue with the Sigil library on net461/netstandard2.0. Now that Sigil has been removed, we no longer need this test case, especially because it is demonstrating some unreliable behavior.

### ProcessExit event handler improvements ([#1332](https://github.com/DataDog/dd-trace-dotnet/pull/1332))
commit 0aaa30eb253a6af8792cbd1ff5918927c663d777
Author: Tony Redondo <tony.redondo@datadoghq.com>
Date:   Mon Apr 5 20:26:00 2021 +0200
    
    * Improves the ProcessExit event handler.
    
    * fixes #if
    
    * Changes from review.
    
    * Update src/Datadog.Trace/Tracer.cs
    
    Co-authored-by: Lucas Pimentel-Ordyna <lucas.pimentel@datadoghq.com>
    
    * Changes from reviews
    
    * Add last delegate logic for CancelKeyPress event.
    
    * Fix test
    
    Co-authored-by: Lucas Pimentel-Ordyna <lucas.pimentel@datadoghq.com>

### Implement Dependabot Config ([#1338](https://github.com/DataDog/dd-trace-dotnet/pull/1338))
commit f0498e505adc9e58ae733c0c3d6566727ff8ffe5
Author: Colin Higgins <colin-higgins@users.noreply.github.com>
Date:   Mon Apr 5 13:56:53 2021 -0400

### Generate fake proj for dependabot ([#1336](https://github.com/DataDog/dd-trace-dotnet/pull/1336))
commit 2ffc01b542fd21f0533d5ac1b8a1c6fa4647eb35
Author: Colin Higgins <colin-higgins@users.noreply.github.com>
Date:   Fri Apr 2 16:14:38 2021 -0400

### Revert "Add a consolidated multi-stage pipeline ([#1320](https://github.com/DataDog/dd-trace-dotnet/pull/1320))" ([#1335](https://github.com/DataDog/dd-trace-dotnet/pull/1335))
commit 45171ff893fce8c3488b40dc168d1d34e37d2ba6 (tag: v1.25.2-prerelease)
Author: Kevin Gosse <krix33@gmail.com>
Date:   Fri Apr 2 21:18:40 2021 +0200

    This reverts commit 74d81fd710a0efb98fcc444e78deaced530dc224.

### Bump version to 1.25.2-prerelease ([#1333](https://github.com/DataDog/dd-trace-dotnet/pull/1333))
commit 6756f19aa9d44052885725d89565f54b9d431b6a
Author: Kevin Gosse <krix33@gmail.com>
Date:   Fri Apr 2 14:02:13 2021 +0200

### Add support for flushing partial traces ([#1313](https://github.com/DataDog/dd-trace-dotnet/pull/1313))
commit e3dfe290a1d9f13f4e007c7ab70c1d26b4394131
Author: Kevin Gosse <krix33@gmail.com>
Date:   Thu Apr 1 19:31:03 2021 +0200
    
    * Implement partial flush

### Bump version to 1.25.1 ([#1330](https://github.com/DataDog/dd-trace-dotnet/pull/1330))
commit b8e639b319485c445edbf6997dba617299398079 (tag: v1.25.1-prerelease)
Author: Kevin Gosse <krix33@gmail.com>
Date:   Thu Apr 1 17:44:41 2021 +0200

### Replace Thread.Sleep with Task.Delay in dogstatsd ([#1326](https://github.com/DataDog/dd-trace-dotnet/pull/1326))
commit 80b37740864dae2cc05dc47db39d5c7ba7ce2eec
Author: Kevin Gosse <krix33@gmail.com>
Date:   Thu Apr 1 10:58:22 2021 +0200

### CODE REVIEW

**Comment by zacharycmontoya** [comment]
@dszmigielski I encourage you to also take the next commit which Datadog is hopeful resolves a flaky unit test: https://github.com/DataDog/dd-trace-dotnet/commit/d62b8728a32c471082ee00f475203ad123e6e76b

File: test/Datadog.Trace.Tests/Tagging/TagsListTests.cs [hunk 1]
```diff
@@ -68,6 +68,39 @@ public void GetAll_When_MissingTags()
             ValidateTags(tags.GetAllTags(), values);
         }
 
+        [Fact]
```
Comment (zacharycmontoya) [inline test/Datadog.Trace.Tests/Tagging/TagsListTests.cs:71]: Changes from this file should be removed in a follow-up commit, as we decided to write our tests differently than OTel

**Comment by dszmigielski** [comment]
> @dszmigielski I encourage you to also take the next commit which Datadog is hopeful resolves a flaky unit test: [DataDog/dd-trace-dotnet@d62b872](https://github.com/DataDog/dd-trace-dotnet/commit/d62b8728a32c471082ee00f475203ad123e6e76b)

I have picked that one additional commit, thanks for pointing that out :)

File: .azure-pipelines/crank.yml [hunk 2]
```diff
@@ -41,10 +41,10 @@ jobs:
       version: 3.1.x
 
   - task: UseDotNet@2
-    displayName: install dotnet core sdk 5.0
+    displayName: install dotnet core sdk 5.0.103
```
Comment (pellared) [inline .azure-pipelines/crank.yml:44]: This was introduced here: https://github.com/DataDog/dd-trace-dotnet/pull/1352
However, this is not needed. IMO a better fix one was already merged in OTel repo: https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/pull/120

@andrewlock @tonyredondo 
The corresponding fix in DD was not merged: https://github.com/DataDog/dd-trace-dotnet/pull/1353

File: honeypot/Datadog.Dependabot.Honeypot.csproj [hunk 3]
```diff
@@ -0,0 +1,26 @@
+<Project Sdk="Microsoft.NET.Sdk.Web">
```
Comment (pellared) [inline honeypot/Datadog.Dependabot.Honeypot.csproj:1]: @zacharycmontoya 
I really miss docs describing how to use it and in general all the tools you have built for managing vendor dependencies.

File: .azure-pipelines/crank.yml (hunk 2 above)
Comment (andrewlock) [reply .azure-pipelines/crank.yml:44]: Yes, there was a bit of discussion around that, will likely be merged soon. 

However, I think the fix introduced in https://github.com/DataDog/dd-trace-dotnet/pull/1352 is still worth keeping. It was the only pipeline using a different SDK version to the others.

Comment (pellared) [reply .azure-pipelines/crank.yml:44]: 👍

File: test/Datadog.Trace.Tests/Tagging/TagsListTests.cs (hunk 1 above)
Comment (dszmigielski) [reply test/Datadog.Trace.Tests/Tagging/TagsListTests.cs]: I have reverted these.

**Review by zacharycmontoya — APPROVED** [review]

**Review by nrcventura — APPROVED** [review]

**Pull request merged**